api/envoy/config/common/matcher/v3/matcher.proto

```diff
@@ -81,6 +81,9 @@ message Matcher {
 
         // A list of predicates to be AND-ed together.
         PredicateList and_matcher = 3;
+
+        // The invert of a predicate
+        Predicate not_matcher = 4;
```

Comment:
@aguinet this makes sense API-wise. Given how simple the API change is, I'd suggest folding the implementation &  tests into this PR. Thanks.

@snowp FYI, how come we didn't include this to start with?

/wait

Reply:
@htuch I'm not sure to fully understand your suggestion. The implementation and tests are already in this PR. Do you want them in a separate commit/PR?

Thanks!

Reply:
I think I didn't include it because I thought the `on_no_match` mechanism would be sufficient, but to get that working in the trivial case you need a noop action to terminate the match path (since we required an on_match for each matcher), which seems like a less than ideal pattern